lib/core/src/Cardano/Pool/Ranking.hs

```diff
@@ -0,0 +1,170 @@
+-- Types
+--------------------------------------------------------------------------------
+
+newtype a >= (b :: Nat) = Bounded { unBounded :: a }
+    deriving (Generic, Eq, Show)
+    deriving newtype Ord
+
+unsafeToBounded :: forall a b. (Ord a, Show a, Num a, KnownNat b) => a -> (a >= b)
+unsafeToBounded x =
+    if x >= b
+    then Bounded x
+    else error $ "unsafeToBounded: " ++ show x ++ " > " ++ show b ++ "does not hold"
+  where
+    b = fromIntegral $ natVal $ Proxy @b
+
+data EpochConstants = EpochConstants
+    { leaderStakeInfluence :: Double >= 0
+      -- ^ a_0
+    , punishmentHarshness :: Double >= 0
+      -- ^ \gamma
+    , optimalNumberOfPools :: Int >= 1
+      -- ^ k
+    , totalRewards :: Lovelace >= 1
+      -- ^ Total rewards in an epoch. "R" in the spec.
+    } deriving (Show, Eq, Generic)
+
+saturatedPoolSize :: EpochConstants -> RelativeStakeOf "pool"
+saturatedPoolSize constants =
+    RelativeStake $ 1 / fromIntegral (unBounded $ optimalNumberOfPools constants)
+
+data Pool = Pool
+    { leaderStake :: RelativeStakeOf "pool leader"
+    , cost :: Lovelace
+    , margin :: Margin
+    , recentAvgPerformance :: Double >= 0
+      -- ^ An already averaged performance-value
+    } deriving (Show, Eq, Generic)
+
+newtype Lovelace = Lovelace { getLovelace :: Double }
+    deriving (Show, Eq)
+    deriving newtype (Ord, Num)
+newtype Margin = Margin { unMargin :: Double }
+    deriving (Show, Eq)
+    deriving newtype Ord
+
+-- The value
+-- 0.01 :: RelativeStakeOf "pool"
+-- would mean that a pool has a stake that is 1% of the total active stake in
+-- the epoch.
+newtype RelativeStakeOf (tag :: Symbol)
+    = RelativeStake { unRelativeStake :: Double }
+    deriving (Eq, Ord, Show, Generic)
+    deriving newtype (Num, Fractional)
+
+--------------------------------------------------------------------------------
+-- Formulas from spec
+--------------------------------------------------------------------------------
+
+desirability
+    :: EpochConstants
+    -> Pool
+    -> Double
+desirability constants pool
+    | f_saturated <= c = 0
+    | otherwise    = (f_saturated - c) * (1 - m)
+  where
+    f_saturated = saturatedPoolRewards constants pool
+    m = unMargin $ margin pool
+    c = getLovelace $ cost pool
+
+-- | Total rewards for a pool if it were saturated.
+--
+-- When a0 = 0 this reduces to just p*R*z0 (tested
+-- by @prop_saturatedPoolRewardsReduces@)
+saturatedPoolRewards :: EpochConstants -> Pool -> Double
+saturatedPoolRewards constants pool =
+    let
+        a0 = unBounded $ leaderStakeInfluence constants
+        z0 = unRelativeStake $ saturatedPoolSize constants
+        s = unRelativeStake $ leaderStake pool
+        _R = getLovelace $ unBounded $ totalRewards constants
+        p = unBounded $ recentAvgPerformance $ pool
+        -- ^ technically \hat{p} in the spec
+    in
+        (p * _R) / (1 + a0)
+        * (z0 + ((min s z0) * a0))
```

Comment:
May I suggest to define the "optimalRewards" at the top-level as:

```hs
optimalReward (R, z0, a0) σ s = 
    R / (1 + a0) * (σ' + s'*a0*(σ' - s'*(z0-σ')/z0) / z0)
  where
    σ' = min σ z0              
    s' = min s z0
```

Where the saturated pool reward can be obtained by replacing `σ` by `z0`:

```hs
f_satured = p_apparent * optimalRewards (R, z0, a0) z0 s
```

Reply:
Sounds good, but maybe not necessary now. Causes minor cascading changes.